// @ngxs/websocket-plugin/esm2015/src/symbols.js
import { InjectionToken } from '@angular/core';
export const NGXS_WEBSOCKET_OPTIONS = new InjectionToken('NGXS_WEBSOCKET_OPTIONS');
export function noop(..._args) {
    return function () { };
}
/**
 * Action to connect to the websocket. Optionally pass a URL.
 */
export class ConnectWebSocket {
    constructor(payload) {
        this.payload = payload;
    }
}
ConnectWebSocket.type = '[WebSocket] Connect';
/**
 * Action triggered when a error ocurrs
 */
export class WebsocketMessageError {
    constructor(payload) {
        this.payload = payload;
    }
}
WebsocketMessageError.type = '[WebSocket] Message Error';
/**
 * Action to disconnect the websocket.
 */
export class DisconnectWebSocket {
}
DisconnectWebSocket.type = '[WebSocket] Disconnect';
/**
 * Action triggered when websocket is connected
 */
export class WebSocketConnected {
}
WebSocketConnected.type = '[WebSocket] Connected';
/**
 * Action triggered when websocket is disconnected
 */
export class WebSocketDisconnected {
}
WebSocketDisconnected.type = '[WebSocket] Disconnected';
/**
 * Action to send to the server.
 */
export class SendWebSocketMessage {
    constructor(payload) {
        this.payload = payload;
    }
}
SendWebSocketMessage.type = '[WebSocket] Send Message';
/**
 * Action dispatched when the user tries to connect if the connection already exists.
 */
export class WebSocketConnectionUpdated {
}
WebSocketConnectionUpdated.type = '[WebSocket] Connection Updated';
/**
 * This error is thrown where there is no `type` (or custom `typeKey`) property
 * on the message that came from the server side socket
 */
export class TypeKeyPropertyMissingError extends Error {
    constructor(typeKey) {
        super(`Property ${typeKey} is missing on the socket message`);
    }
}
